File: TUnit.TestProject/AssemblyHooks.cs
namespace TUnit.TestProject;

public abstract class AssemblyHooks
{
    private static int _beforeHook1Calls;

    [Before(Assembly)]
    public static void BeforeHook1()
    {
        _beforeHook1Calls++;
    }


#if NET
    [Before(Assembly)]
    public static async Task BeforeHook2(AssemblyHookContext context)
    {
        await Assert.That(context.TestCount).IsPositive();
    }
#endif

#if NET
    [Before(Assembly), Timeout(30_000)]
    public static async Task BeforeHook4(AssemblyHookContext context, CancellationToken cancellationToken)
    {
        await Assert.That(context.TestCount).IsPositive();
    }
#endif

    [After(Assembly)]
    public static async Task AfterHook1()
    {
        await Assert.That(_beforeHook1Calls).IsEqualTo(1);
    }

#if NET
    [After(Assembly)]
    public static async Task AfterHook2(AssemblyHookContext context)
    {
        await Assert.That(context.TestCount).IsPositive();
    }
#endif

#if NET
    [After(Assembly), Timeout(30_000)]
    public static async Task AfterHook4(AssemblyHookContext context, CancellationToken cancellationToken)
    {
        await Assert.That(context.TestCount).IsPositive();
    }
#endif
}

File: TUnit.TestProject.FSharp/AssemblyHooks.fs
namespace TUnit.TestProject.FSharp

open System.Threading
open System.Threading.Tasks
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Core
open TUnit.Assertions.FSharp.Operations

// F# equivalent of AssemblyHooks.cs

type AssemblyHooks() =
    static let mutable beforeHook1Calls = 0

    [<Before(HookType.Assembly)>]
    static member BeforeHook1() =
        beforeHook1Calls <- beforeHook1Calls + 1

#if NET
    [<Before(HookType.Assembly)>]
    static member BeforeHook2(context: AssemblyHookContext) = async{
        do! check(Assert.That(context.TestCount).IsPositive())
    }
#endif

    [<Before(HookType.Assembly)>]
    [<Timeout(30000)>]
    static member BeforeHook3(cancellationToken: CancellationToken) =
        ()

#if NET
    [<Before(HookType.Assembly)>]
    [<Timeout(30000)>]
    static member BeforeHook4(context: AssemblyHookContext, cancellationToken: CancellationToken) = async {
        do! check(Assert.That(context.TestCount).IsPositive())
    }
#endif

    [<After(HookType.Assembly)>]
    static member AfterHook1() = async {
        do! check(Assert.That(beforeHook1Calls).IsEqualTo(1))
    }

#if NET
    [<After(HookType.Assembly)>]
    static member AfterHook2(context: AssemblyHookContext) = async {
        do! check(Assert.That(context.TestCount).IsPositive())
    }
#endif

    [<After(HookType.Assembly)>]
    [<Timeout(30000)>]
    static member AfterHook3(cancellationToken: CancellationToken) =
        ()

#if NET
    [<After(HookType.Assembly)>]
    [<Timeout(30000)>]
    static member AfterHook4(context: AssemblyHookContext, cancellationToken: CancellationToken)  = async {
        do! check(Assert.That(context.TestCount).IsPositive())
    }
#endif
